Cruise Control blog post (#163)
strimzi/strimzi.github.io: _posts/2020-06-05-cruise-control.md

---
layout: post
title:  "Cluster balancing with Cruise Control"
date: 2020-06-04
author: tom_cooper
---

> Kafka allows you to scale your broker infrastructure to cope with massive load, brokers can be added easily to increase capacity. However, new brokers will only play host to new topics partitions created after the broker was added to the cluster. If certain brokers in your existing cluster are currently under heavy load, then adding a new broker to the cluster will not relieve pressure on the others. To do that you would need to [reassign partiton replicas](https://kafka.apache.org/documentation/#basic_ops_partitionassignment) to the new broker. Similarly, imagine that you have a cluster where, for whatever reason, one broker happens to hold a disproportionate number of partition leaders for partitions under heavy load. To solve this issue you would need to figure out which partition are under load, identify other brokers with lower load which hold replicas for that partition, and then reassign the partition leadership to even out the load across the cluster. This is no easy task. 

Reviewer: ```suggestion
Kafka allows you to scale your broker infrastructure to cope with massive load, brokers can be added easily to increase capacity. However, new brokers will only play host to new topics partitions created after the broker was added to the cluster. If certain brokers in your existing cluster are currently under heavy load, then adding a new broker to the cluster will not relieve pressure on the others. To do that you would need to [reassign partiton replicas](https://kafka.apache.org/documentation/#basic_ops_partitionassignment) to the new broker. Similarly, imagine that you have a cluster where, for whatever reason, one broker happens to hold a disproportionate number of partition leaders for partitions under heavy load. To solve this issue you would need to figure out which partitions are under load, identify other brokers with lower load which hold replicas for those partitions, and then reassign the partition leadership to even out the load across the cluster. This is no easy task. 
```

Reviewer: TLDR; 

```which partition are``` -> ```which partitions are```
```that partition``` -> ```those partitions```

Enter [Cruise Control](https://github.com/linkedin/cruise-control), a project from Linkedin, which automates the _balancing_ of load across a Kafka cluster. [Last year]({% post_url 2019-08-08-hacking-for-cruise-control %}), Kyle Liberti described how you could manually get Cruise Control working with Strimzi. With the latest [0.18 release](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.18.0), Strimzi now provides native Cruise Control support. 

<!--more-->

## Cruise Control overview

Cruise Control started out as an internal project at Linkedin and was used for managing their in-house Kafka clusters. It was open sourced in 2016 and has seen constant development since then. 

> Cruise Control consists of a main server implementation that contains several sub-systems for monitoring, analyzing and proposing changes to the Kafka Cluster it monitors. As well as the main server, Cruise Control provides a Kafka metrics reporter implementation that, once installed into the Kafka brokers, records a wide range of metrics in addition to the default metrics provided by the brokers themselves.

Reviewer: From what I remember, the metrics the metric reporter records are from the Kafka brokers with the exception the the CPU util metric, which is taken by the metric reporter itself. For that reason I think we should reword the last phrase/sentence to something like this:
```suggestion
Cruise Control consists of a main server implementation that contains several sub-systems for monitoring, analyzing and proposing changes to the Kafka Cluster it monitors. As well as the main server, Cruise Control provides a Kafka metrics reporter implementation that, once installed into the Kafka brokers, filters and records metrics provide by the brokers themselves for further analysis by the mainl server.
```
to avoid the phrase ```records a wide range of metrics in addition```

> The Cruise Control server periodically collects the metrics recorded by the metrics reporters (which are sent to topics within the cluster) to infer the traffic pattern of each partition. Based on the traffic characteristics and distribution of all the partitions, it derives the load impact of each partition on the brokers. Cruise Control then builds a workload model to simulate the workload of the Kafka cluster. 

Reviewer 2: There are a couple of configuration, on the metrics reporter side to specify how often it has to write metrics on the metrics topic; on the CC side to specify how often the metrics sampler has to read from the topic.
@tomncooper "which are sent to topics within the cluster" ... it too generic, the metrics reporter sends to a specific "metrics topic".

> The user can define a list of _optimization goals_. These goals each define a set of parameter values that the cluster state should meet. For example a goal can enforce that network traffic for each broker should not differ by more than +/- 10% of the cluster's average. These goals are configured in a priority order and when required the Cruise Control server's analyzer sub-system will attempt to change the workload model to meet each goal. These changes consist of switching the lead replica of a given partition replica leadership or reassigning partitions to other brokers. The analyzer will descend down the list of goals and first check if the current cluster configuration meets that goal and if it doesn't will attempt to alter the cluster model until it does. At each step any generated changes are passed to all the previous goals in the list to check if they will then violate those higher priority goals.

Reviewer: ```given partition replica leadership``` -> ```given partition```

> Each of the goals can either be _hard_ or _soft_. A hard goal **must** be satisfied, if the analyzer proposes changes that violate a hard goal then those changes are rejected. A soft goal can be violated. Therefore an _optimization proposal_, which is a collection of changes resulting from a full pass over a given list of goals, will meet the requirements of all the hard goals in that list but may not satisfy all the soft goals. It is worth noting that if it is not possible to meet all the hard goals then no optimization proposal will be produced.

Reviewer: ```all of the soft goal```

Not completely sure if the "of" is needed here but my brain inserts it when I read this line, do what you will!

For a more detailed breakdown of how Cruise Control does what it does, see the project's [wiki](https://github.com/linkedin/cruise-control/wiki/Overview), this introduction talk ([video](https://www.youtube.com/watch?v=lf31udm9cYY)/[slides](https://www.slideshare.net/JiangjieQin/introduction-to-kafka-cruise-control-68180931)) and this [updated talk](https://www.youtube.com/watch?v=jdo6F21gI8g) from the developers themselves.

## Strimzi Cruise Control support

> Cruise Control's optimization approach allows it to perform several different operations linked to balancing cluster load. As well as rebalancing a whole cluster, Cruise Control also has the ability to do this balancing automatically when adding and removing brokers or changing topic replica values. Cruise Control can also supply information on cluster load, can monitor the cluster for anomalies and even perform (limited) automated self-healing actions. 

Reviewer: ```suggestion
Cruise Control's optimization approach allows it to perform several different operations linked to balancing cluster load. As well as rebalancing a whole cluster, Cruise Control also has the ability to do this balancing automatically when adding and removing brokers or changing topic replica values. Cruise Control can also supply information on cluster load, monitor the cluster for anomalies, and even perform (limited) automated self-healing actions. 
```

### Getting an optimization proposal

Now you have Cruise Control deployed you can ask it to generate a optimization proposal. To do this you will need to apply a `KafkaRebalance` custom resource (the definition for this is installed when you install Strimzi). A basic `KafkaRebalance` is shown below:

```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
  spec: {}
```

And you can deploy this like any other resource:

```bash
kubectl apply -f kafka-rebalance.yaml -n myproject
```

This rebalance has an empty `spec` field and so will use the default rebalance settings. This means it will use the `default.goals` (mentioned above) and will return the most recent _cached optimization proposal_. Once you apply this resource to your Kafka cluster the Cluster Operator will issue the relevant requests to Cruise Control to fetch the optimization proposal. The Cluster Operator will keep the status of your `KafkaRebalance` updated with the progress from Cruise Control. You can check this using:

```bash

> $ kubectl describe kafkarebalance -n myproject my-rebalance      

Reviewer 2: I would put the namespace parameter at the end. I think having `kafkarebalance my-rebalance` is more readable.

#### Interpreting an optimization proposal

In the case shown above, Cruise Control has suggested 55 partition replicas (totalling 12MB of data) should be move between separate brokers (_inter_-broker) in the cluster. This will involve moving data across the network, but 12MB is not a lot so this should not have too much of an impact. If this number was very high you may want to time this rebalance for a period of less load, as moving large amounts of data could effect the performance of the cluster as a whole. 

There are no _intra_ (between disks on the same broker) movements proposed. These kind of movements only involve movement between disks so have no network impact, but may effect the performance the particular brokers whose disks are exchanging data. 

> Finally, this proposal is suggesting 24 leader replica changes. This is the cheapest change that can be applied as it only involves a configuration change in Zookeeper.

Reviewer 2: ZooKeeper

### Starting a cluster rebalance

> If you are happy with the proposal then you can apply the `approve` annotation to the `KafkaRebalance` resource to tell Cruise Control to start applying the changes:

Reviewer 2: I would say "the approve value to the strimzi.io/rebalance annotation". Wdyt? Or something like that.

```bash
$ kubectl annotate kafkarebalance my-rebalance strimzi.io/rebalance=approve
```

> It is important to note that if you wait a long time between asking for a proposal and approving it, the underlying cluster might be in a very different state to the time when you requested the proposal. In this case it is a good idea to _refresh_ the proposal, just to make sure it is up-to-date. You can do this by attaching the `refresh` annotation to the `KafkaRebalance` resource:

Reviewer 2: ditto as above.

```bash
$ kubectl annotate kafkarebalance my-rebalance strimzi.io/rebalance=refresh
```

> Once you apply the `approve` annotation Cruise Control will begin enacting the changes from that proposal. The Cluster Operator will keep the status of the `KafkaRebalance` resource up-to-date with the progress of the rebalance whilst it is ongoing. You can check this by running:

Reviewer 2: or thinking more you could say `strimzi.io/rebalance=approve` annotation

```bash
$ kubectl describe kafkarebalance -n myproject my-rebalance      
```

> Whilst the rebalance is in progress the Status will show `Rebalancing` and once it is finished it will show `Ready`.

Reviewer 2: no snippet example for this?

### Custom rebalance goals

In the `KafkaRebalance` example above we didn't specify any additional configurations, in that case Cruise Control generates the optimization proposal (or returns the cached proposal if it is ready) using the configured `default.goals`. It is also possible to specify a list of custom goals for each individual rebalance. This is so that you can request a rebalance focused on a specific set of goals, for example just making sure that CPU load is balanced. You can specify the goals in the `KafkaRebalance` resource:

```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  goals:
    - NetworkInboundCapacityGoal
    - DiskCapacityGoal
    - RackAwareGoal
    - NetworkOutboundCapacityGoal
    - CpuCapacityGoal
    - ReplicaCapacityGoal
```

When you ask for a rebalance with custom goals, Cruise Control will first check if the current cached optimization proposal already meets the list of custom goals you specified. If so, you will get your proposal back quickly. If not, you may have to wait a while for Cruise Control to perform the calculations.

> One thing to watch out for here, is that your list of custom goals contains all the configured `hard.goals` (the default `hard.goals` are listed in [the docs](https://strimzi.io/docs/operators/latest/using.html#con-optimization-goals-str)). If it doesn't then Cruise Control will throw an error, after all why would say these goals are _hard_ (really important) and then ignore them? But, if you really do only care about a few goals for this one off rebalance, you can tell Cruise Control to relax by adding `skipHardGoalCheck: true` to the `KafkaRebalance` resource.

Reviewer: ```suggestion
One thing to watch out for here, is that your list of custom goals contains all the configured `hard.goals` (the default `hard.goals` are listed in [the docs](https://strimzi.io/docs/operators/latest/using.html#con-optimization-goals-str)). If it doesn't then Cruise Control will throw an error, after all why you would say these goals are _hard_ (really important) and then ignore them? But, if you really do only care about a few goals for this one off rebalance, you can tell Cruise Control to relax by adding `skipHardGoalCheck: true` to the `KafkaRebalance` resource.
```

### Stopping a rebalance

> Rebalances can take a long time and may impact the performance of your cluster, therefore you may want to stop a rebalance if that is causing an issue. It is also possible that you may approve an optimization proposal that you later realise was not what you wanted. In either case, you need to be able to stop an ongoing rebalance. To do this you can attach the `stop` annotation to the `KafkaRebalance` resource at any time:

Reviewer 3: attach annotation sounds weird ... why not just say annotate?

```bash
$ kubectl annotate kafkarebalance my-rebalance strimzi.io/rebalance=stop
```

> The status of the resource should then show `Stopped`. It is worth baring in mind that Cruise Control batches changes to be passed to the Kafka Cluster, so whilst it will stop sending changes as soon as you attach the annotation, the Kafka Cluster may still be processing the last batch of changes.

Reviewer: ```baring in mind```  -> ```bearing in mind```

May be a British versus American English thing though!

---
layout: post
title:  "Cluster balancing with Cruise Control"

> date: 2020-06-04

Reviewer 2: wrong date here
